google/gax/bundling.py

```diff
@@ -0,0 +1,252 @@
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+#     * Redistributions of source code must retain the above copyright
+# notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above
+# copyright notice, this list of conditions and the following disclaimer
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""Provides behavior that supports request bundling."""
+
+from __future__ import absolute_import
+
+import collections
+import threading
+
+
+def _str_dotted_getattr(obj, name):
+    """Expands extends getattr to allow dots in x to indicate nested objects.
+
+    Args:
+       obj: an object
+       name: a name for a field in the object
+
+    Returns:
+       the value of named attribute
+
+    Raises:
+       AttributeError if the named attribute does not exist
+    """
+    if name.find('.') == -1:
+        return getattr(obj, name)
+    for part in name.split('.'):
+        obj = getattr(obj, part)
+    return str(obj) if obj is not None else None
+
+
+def compute_bundle_id(obj, descriminator_fields):
+    """Computes a bundle id from the descriminator fields of `obj`.
+
+    descriminator_fields may include '.' as a separator, which is used to
+    indicate object traversal.  This is meant to allow fields in the
+    computed bundle_id.
+
+    the id is a tuple computed by going through the descriminator fields in
+    order and obtaining the str(value) object field (or nested object field)
+
+    if any descriminator field cannot be found, ValueError is raised.
+
+    Args:
+      obj: an object
+      descriminator_fields: a list of descriminator fields in the order to be
+        to be used in the id
+
+    Returns:
+      tuple: computed as described above
+
+    Raises:
+      AttributeError: if any descriminator fields attribute does not exist
```

Comment:
So what happens if, say, the logging API is configured to bundle by log name and labels, but I want to write a log with no labels? Is this impossible if bundling is enabled?